### ISSUE TITLE
RAM usage by converting from sparse to dense

### ISSUE THREAD

**Issue by chbeltz** [Teichlab/celltypist#29]
Is there a reason why the input data is converted to an np.array rather than accepting sparse matrices when running .train? Skimming the remainder of the code, I cannot seem to find anything that would not also work with sparse matrices. The reason I am asking is that this conversion to an array seems to be the reasons I find myself running out of  RAM quite frequently when working with larger datasets.

Thanks

**Comment by prete** [comment]
Hi @chbeltz, thank you for using CellTypist! 

I think the issue you're seeing has to do with the scaling technique that's used. Have a look at [classifier.py > celltype](https://github.com/Teichlab/celltypist/blob/e48a998d5fc6df7c3570b38c4bd3605b12144f59/celltypist/classifier.py#L352), particularly at this:
```python
        logger.info(f"⚖️ Scaling input data")
        means_ = self.model.scaler.mean_[lr_idx]
        sds_ = self.model.scaler.scale_[lr_idx]
        self.indata = (self.indata[:, k_x_idx] - means_) / sds_
        self.indata[self.indata > 10] = 10
```

Unfortunately, `self.indata[:, k_x_idx] - means_` densifies the matrix, possibly leading to you running out of RAM with large datasets.

**Comment by chbeltz** [comment]
Ah, I wasn't aware that sparse matrices were densified upon substraction by a vector. That's unfortunate.

Thanks!

**Comment by chbeltz** [comment]
Can you imagine implenting a use_sparse switch that preserves sparsity by skipping the subtraction by the mean during scaling? Same principle as the with_mean option of sklearn.preprocessing.StandardScaler.

**Comment by ChuanXu1** [comment]
@chbeltz, for SGD logistic regression, according to the sklearn package

> For best results using the default learning rate schedule, the data should have zero mean and unit variance.

Skipping the subtraction by the mean may not be a good practice, wdyt?

**Comment by chbeltz** [comment]
@ChuanXu1 I have not been able to find a lot of empirical data on the effect of non-zero centered distributions of the input onto the performance of SGD, so I'm having a hard time weighing the pros and cons. However, I feel that if the alternative is that people with limited computing resources may decide to not use the software at all, it may be preferable to provide an option that may lead to less than optimal results, but to results nonetheless.

**Comment by ChuanXu1** [comment]
@chbeltz, it sounds reasonable. I added these changes (a `with_mean` paramter in [celltypist.train](https://celltypist.readthedocs.io/en/latest/celltypist.html#celltypist.train.train)) to optimize the RAM usage during training at a possible cost of reduced performance dfb11e05e3e95fb4906a24ae4f890988eba13031 

This parameter will be available at (next version of) CellTypist. Thx!

**Comment by chbeltz** [comment]
Much appreciated, thank you!!
